### PR TITLE
Channel access for Stations with scheduled access period (SP or CBAP)

### PR DESCRIPTION
**Pull request by tommyazz**
Stations with allocated SP or CBAP might be prevented from competing for channel access during a broadcast CBAP

### CODE REVIEW

File: src/wifi/model/dmg-wifi-mac.h [hunk 1]
```diff
@@ -243,6 +243,17 @@ class DmgWifiMac : public RegularWifiMac
...
   virtual void Configure80211ad (void);
 
+  /**
+   * This function registers the allocation requests accepted by the PCP/AP.
+   * \param info The DMG allocation info field associated with the request.
+   */
+  void RegisterAllocatedRequest (const DmgAllocationInfo &info);
+  /**
+   * Get whether a Station can compete for channel access during a broadcast CBAP.
+   * \param staAid The AID of the Station.
```
Comment (mattia-lecci) [inline src/wifi/model/dmg-wifi-mac.h]:
```suggestion
   * \param staAid The AID of the target STA
```

File: src/wifi/model/dmg-wifi-mac.h [hunk 2]
```diff
@@ -765,6 +777,16 @@ class DmgWifiMac : public RegularWifiMac
...
 
+  typedef struct 
+    {
+      AllocationID id;
+      uint16_t dstAid;
+    } AllocatedDataStruct;
+
+  typedef std::vector<AllocatedDataStruct> AllocatedRequestsVec;
+  AllocatedRequestsVec m_allocatedRequests;     //!< Vector of requested allocations granted to this STA
+  bool m_accessInCbap;                          //!< Flag to indicate whether a Station with allocated SP/CBAP is allowed to compete in a CBAP.
```
Comment (mattia-lecci) [inline src/wifi/model/dmg-wifi-mac.h]: keep a consistent naming style
```suggestion
  bool m_accessInCbap;                          //!< Flag to indicate whether a STA with allocated SP/CBAP is allowed to compete in a CBAP.
```

File: src/wifi/model/dmg-wifi-mac.cc [hunk 3]
```diff
@@ -91,6 +91,13 @@ DmgWifiMac::GetTypeId (void)
                     MakeEnumChecker (RELAY_FD_AF, "Full Duplex",
                                      RELAY_HD_DF, "Half Duplex",
                                      RELAY_BOTH, "Both"))
+    
+    /* Channel access in the DTI */
+    .AddAttribute ("AccessInCbap", "Whether a Station with allocated SP or CBAP"
```
Comment (mattia-lecci) [inline src/wifi/model/dmg-wifi-mac.cc]: misleading attribute name. maybe AccessCbapIfAllocated?

File: src/wifi/model/dmg-wifi-mac.cc [hunk 4]
```diff
@@ -91,6 +91,13 @@ DmgWifiMac::GetTypeId (void)
...
+                   "is allowed to compete for channel access during a broadcast CBAP.",
```
Comment (mattia-lecci) [inline src/wifi/model/dmg-wifi-mac.cc]: missing space in new line

File: src/wifi/model/dmg-ap-wifi-mac.cc [hunk 5]
```diff
@@ -1301,7 +1301,9 @@ DmgApWifiMac::StartDataTransmissionInterval (void)
                 }
             }
           else if ((field.GetAllocationType () == CBAP_ALLOCATION) &&
-                  ((field.GetSourceAid () == AID_BROADCAST) || (field.GetSourceAid () == AID_AP) || (field.GetDestinationAid () == AID_AP)))
+                  (((field.GetSourceAid () == AID_BROADCAST) && AccessAllowedInBroadcastCbap (AID_AP)) ||
```
Comment (mattia-lecci) [inline src/wifi/model/dmg-ap-wifi-mac.cc:1304]: could you explain this?

Comment (tommyazz) [reply src/wifi/model/dmg-ap-wifi-mac.cc:1304]: Basically, in this part of `DmgApWifiMac::StartDataTransmissionInterval` (same for `DmgStaWifiMac::StartDataTransmissionInterval`) if the allocation is a CBAP, an STA can access the contention period if it's destination or source of the CBAP or the CBAP is broadcast (i.e. the source AID is the broadcast AID). In the latter case, I added the check on the accessibility of the broadcast CBAP by the STA.

**Review by mattia-lecci — APPROVED** [review]

**Review by MatteoDrago — APPROVED** [review]

**Review by mattia-lecci — APPROVED** [review]

**Review by MatteoDrago — APPROVED** [review]

**Pull request merged**